test/framework/controlplane_helpers.go

```diff
@@ -310,7 +311,9 @@ func UpgradeControlPlaneAndWaitForUpgrade(ctx context.Context, input UpgradeCont
 	Expect(err).ToNot(HaveOccurred())
 
 	input.ControlPlane.Spec.Version = input.KubernetesUpgradeVersion
-
+	if input.KubernetesVersionUpgradeMachineTemplateTo != nil {
+		input.ControlPlane.Spec.MachineTemplate.InfrastructureRef.Name = pointer.StringDeref(input.KubernetesVersionUpgradeMachineTemplateTo, "")
```

Comment:
```suggestion
		input.ControlPlane.Spec.MachineTemplate.InfrastructureRef.Name = *input.KubernetesVersionUpgradeMachineTemplateTo
```
nit. Should be fine as we already handle the nil one line above

Reply:
same in `UpgradeMachineDeploymentsAndWait`